akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -537,35 +570,60 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
   override def preStart(): Unit = {
     tryInit(_initial)
     if (activeInterpreters.isEmpty) context.stop(self)
+    else if (shortCircuitBuffer != null) shortCircuitBatch()
+  }
+
+  private def shortCircuitBatch(): Unit = {
+    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && activeInterpreters.nonEmpty)
+      shortCircuitBuffer.poll() match {
+        case b: BoundaryEvent ⇒ processEvent(b)
+        case Resume           ⇒ finishShellRegistration()
+      }
+    if (!shortCircuitBuffer.isEmpty && currentLimit == 0 && !resumeScheduled) self ! Resume
+  }
+
+  private def processEvent(b: BoundaryEvent): Unit = {
+    val shell = b.shell
+    if (!shell.isTerminated && (shell.isInitialized || tryInit(shell))) {
+      currentLimit = shell.receive(b, currentLimit)
+      if (shell.isTerminated) {
+        activeInterpreters -= shell
+        if (activeInterpreters.isEmpty && newShells.isEmpty) context.stop(self)
+      }
+    }
+    resumeScheduled = shell.resumeScheduled
   }
 
   override def receive: Receive = {
     case b: BoundaryEvent ⇒
-      val shell = b.shell
-      if (!shell.isTerminated && (shell.isInitialized || tryInit(shell))) {
-        shell.receive(b)
-        if (shell.isTerminated) {
-          activeInterpreters -= shell
-          if (activeInterpreters.isEmpty && newShells.isEmpty) context.stop(self)
-        }
-      }
-    case Resume ⇒ finishShellRegistration()
+      resumeScheduled = false
```

Comment:
Idea here was to add logic to not send additional Resume message (to continue processing short circuit queue if limit is 0 but there are some events in short circuit queue) when shell already sent Resume. 

I checked that it works without this logic - so will remove it